app/src/main/java/org/mozilla/fenix/compose/cfr/CFRPopup.kt

```diff
@@ -0,0 +1,610 @@
+ * @property indicatorArrowStartOffset Maximum distance between the popup start and the indicator arrow.
+ * If there isn't enough space this could automatically be overridden up to 0.
+ * @property indicatorArrowHeight How tall the indicator arrow should be.
+ * This will also affect how wide the base of the indicator arrow will be.
+ * @property cornerRadius Corner radius for the popup body.
+ * @property elevation The z-coordinate at which to place the popup.
+ * This controls the size of the shadow below the popup.
+ */
+data class CFRPopupProperties(
+    val indicatorDirection: IndicatorDirection = IndicatorDirection.UP,
+    val dismissOnBackPress: Boolean = true,
+    val dismissOnClickOutside: Boolean = true,
+    val overlapAnchor: Boolean = false,
+    val indicatorArrowStartOffset: Dp = 30.dp,
+    val indicatorArrowHeight: Dp = 10.dp,
+    val cornerRadius: Dp = 10.dp,
+    val elevation: Dp = 10.dp,
+)
+
+/**
+ * CFR - Contextual Feature Recommendation popup.
+ *
+ * @param text [String] shown as the popup content.
+ * @param anchor [View] that will serve as the anchor of the popup and serve as lifecycle owner
+ * for this popup also.
+ * @param properties [CFRPopupProperties] allowing to customize the popup behavior.
+ * @param onDismiss Callback for when the popup is dismissed indicating also if the dismissal
+ * was explicit - by tapping the "X" button or not.
+ * @param action Optional other composable to show just below the popup text.
+ */
+class CFRPopup(
+    private val text: String,
+    private val anchor: View,
+    private val properties: CFRPopupProperties = CFRPopupProperties(),
+    private val onDismiss: (Boolean) -> Unit = {},
+    private val action: @Composable (() -> Unit)? = null
+) {
+    // This is just a facade for the CFRPopupFullScreenLayout composable offering a cleaner API.
+
+    @VisibleForTesting
+    internal var popup: WeakReference<CFRPopupFullScreenLayout>? = null
+
+    /**
+     * Construct and display a styled CFR popup shown at the coordinates of [anchor].
+     * This popup will be dismissed when the user clicks on the "x" button or based on other user actions
+     * with such behavior set in [CFRPopupProperties].
+     */
+    fun show() {
+        anchor.post {
+            CFRPopupFullScreenLayout(text, anchor, properties, onDismiss, action).apply {
+                this.show()
+                popup = WeakReference(this)
+            }
+        }
+    }
+
+    /**
+     * Immediately dismiss this CFR popup.
+     * The [onDismiss] callback won't be fired.
+     */
+    fun dismiss() {
+        popup?.get()?.dismiss()
+    }
+}
+
+/**
+ * [AbstractComposeView] that can be added or removed dynamically in the current window to display
+ * a [Composable] based popup anywhere on the screen.
+ */
+@SuppressLint("ViewConstructor") // Intended to be used only in code, don't need a View constructor
+@VisibleForTesting
+internal class CFRPopupFullScreenLayout(
```

Comment:
This file is pretty large. Would it make sense to have `CFRPopupFullScreenLayout` and `CFRPopupContent` in their own files?

Reply:
Also thought about this but kept it here to better encapsulate it as it should not be used by other classes.
But the same can be said about `CFRPopupShape`.
I'll extract it to a new file, thanks!